File: python_src/tools/namespace_route.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .common import run_command, text_response, ToolResponse, require_ident


def manage_namespace_route(args: dict[str, Any]) -> ToolResponse:
    operation = args.get("operation")
    release_name = args.get("releaseName")
    route = args.get("route")
    working_directory = args.get("workingDirectory") or str(Path.cwd())

    try:
        namespace = require_ident(args.get("namespace"), "namespace")
        if release_name:
            release_name = require_ident(release_name, "releaseName")
        if route:
            route = require_ident(route, "route")

        if operation == "create":
            return _create_namespace_route(namespace, release_name, route, working_directory)
        if operation == "delete":
            return _delete_namespace_route(namespace)
        if operation == "check":
            return _check_namespace_route(namespace)

        raise ValueError(f"Unknown operation: {operation}")
    except Exception as error:  # noqa: BLE001
        return text_response(f"Namespace/route operation failed: {error}", is_error=True)


def _create_namespace_route(
    namespace: str, release_name: str | None, route: str | None, working_dir: str
) -> ToolResponse:
    command = ["./create-namespace-route.sh", namespace]
    if release_name:
        command.append(release_name)
    if route:
        command.append(route)

    stdout, stderr = run_command(command, cwd=working_dir, max_bytes=5 * 1024 * 1024)

    response = "# Namespace and Route Creation\n\n"
    response += f"**Namespace:** {namespace}\n"
    if release_name:
        response += f"**Release Name:** {release_name}\n"
    if route:
        response += f"**Route:** {route}\n\n"

    response += f"## Output\n\n```\n{stdout}```\n"
    if stderr:
        response += f"\n## Warnings\n\n```\n{stderr}```\n"

    return text_response(response)


def _delete_namespace_route(namespace: str) -> ToolResponse:
    try:
        stdout, _ = run_command(["oc", "delete", "namespace", namespace])
        response = "# Namespace Deletion\n\n"
        response += f"**Namespace:** {namespace}\n\n"
        response += "Namespace deleted successfully\n\n"
        response += f"```\n{stdout}```\n"
        return text_response(response)
    except Exception:  # noqa: BLE001
        return text_response(
            f"Failed to delete namespace {namespace}. It may not exist or there may be resources preventing deletion.",
            is_error=True,
        )


def _check_namespace_route(namespace: str) -> ToolResponse:
    response = "# Namespace and Route Status\n\n"
    response += f"**Namespace:** {namespace}\n\n"

    try:
        ns_status, _ = run_command(["oc", "get", "namespace", namespace])
        response += f"## Namespace Status\n\n```\n{ns_status}```\n\n"
    except Exception:  # noqa: BLE001
        response += "## Namespace Status\n\nNamespace does not exist\n\n"
        return text_response(response, is_error=True)

    try:
        route_status, _ = run_command(["oc", "get", "routes", "-n", namespace])
        response += f"## Routes\n\n```\n{route_status}```\n\n"
    except Exception:  # noqa: BLE001
        response += "## Routes\n\nNo routes found or unable to retrieve\n\n"

    return text_response(response)

File: src/tools/namespace-route.ts
import { exec } from "child_process";
import { promisify } from "util";
import * as path from "path";

const execAsync = promisify(exec);

interface ManageNamespaceRouteArgs {
  operation: "create" | "delete" | "check";
  namespace: string;
  releaseName?: string;
  route?: string;
  workingDirectory?: string;
}

export async function manageNamespaceRoute(args: any) {
  const {
    operation,
    namespace,
    releaseName,
    route,
    workingDirectory = process.cwd(),
  } = args as ManageNamespaceRouteArgs;

  try {
    switch (operation) {
      case "create":
        return await createNamespaceRoute(namespace, releaseName, route, workingDirectory);
      
      case "delete":
        return await deleteNamespaceRoute(namespace);
      
      case "check":
        return await checkNamespaceRoute(namespace);
      
      default:
        throw new Error(`Unknown operation: ${operation}`);
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      content: [
        {
          type: "text",
          text: `Namespace/route operation failed: ${errorMessage}`,
        },
      ],
      isError: true,
    };
  }
}

async function createNamespaceRoute(
  namespace: string,
  releaseName: string | undefined,
  route: string | undefined,
  workingDir: string
) {
  let command = `cd "${workingDir}" && ./create-namespace-route.sh ${namespace}`;
  
  if (releaseName) {
    command += ` ${releaseName}`;
  }
  
  if (route) {
    command += ` ${route}`;
  }

  const { stdout, stderr } = await execAsync(command, {
    cwd: workingDir,
    maxBuffer: 1024 * 1024 * 5,
  });

  let response = `# Namespace and Route Creation\n\n`;
  response += `**Namespace:** ${namespace}\n`;
  if (releaseName) response += `**Release Name:** ${releaseName}\n`;
  if (route) response += `**Route:** ${route}\n\n`;
  
  response += `## Output\n\n\`\`\`\n${stdout}\`\`\`\n`;
  
  if (stderr) {
    response += `\n## Warnings\n\n\`\`\`\n${stderr}\`\`\`\n`;
  }

  return {
    content: [{ type: "text", text: response }],
  };
}

async function deleteNamespaceRoute(namespace: string) {
  const command = `oc delete namespace ${namespace}`;

  try {
    const { stdout } = await execAsync(command);
    
    let response = `# Namespace Deletion\n\n`;
    response += `**Namespace:** ${namespace}\n\n`;
    response += `✅ Namespace deleted successfully\n\n`;
    response += `\`\`\`\n${stdout}\`\`\`\n`;

    return {
      content: [{ type: "text", text: response }],
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text",
          text: `Failed to delete namespace ${namespace}. It may not exist or there may be resources preventing deletion.`,
        },
      ],
      isError: true,
    };
  }
}

async function checkNamespaceRoute(namespace: string) {
  let response = `# Namespace and Route Status\n\n`;
  response += `**Namespace:** ${namespace}\n\n`;

  // Check namespace
  try {
    const { stdout: nsStatus } = await execAsync(`oc get namespace ${namespace}`);
    response += `## Namespace Status\n\n\`\`\`\n${nsStatus}\`\`\`\n\n`;
  } catch (error) {
    response += `## Namespace Status\n\n❌ Namespace does not exist\n\n`;
    return {
      content: [{ type: "text", text: response }],
      isError: true,
    };
  }

  // Check routes in namespace
  try {
    const { stdout: routeStatus } = await execAsync(`oc get routes -n ${namespace}`);
    response += `## Routes\n\n\`\`\`\n${routeStatus}\`\`\`\n\n`;
  } catch (error) {
    response += `## Routes\n\n❌ No routes found or unable to retrieve\n\n`;
  }

  return {
    content: [{ type: "text", text: response }],
  };
}
